### ISSUE TITLE
Improve event dispatch mechanism

### ISSUE THREAD

**Issue by mentlerd** [DV8FromTheWorld/JDA#417]
The current ListenerAdapter is usable, but definitely not elegant, and does not allow listener objects to be arbitary subclasses.
If I am not wrong, the source compatibility is Java 8 already, so interface defaults could be used to improve the current state significantly.

My proposal would be a TypedEventListener interface, with the same signatures as ListenerAdapter, but with empty default methods, and a visitor pattern for the Event class.

Backwards compatibility could be implemented by simply making ListenerAdapter implement TypedEventListener, so there would be no duplicate code in the event dispatch mechanism.

I am willing to implement this myself, but I decided I would ask for some developer input beforehand, so my efforts would not be in vain.

**Comment by meowingtwurtle** [comment]
What do you mean `does not allow listener objects to be arbitrary subclasses`? ListenerAdapter is just a partial implementation of EventListener, which is registered with InterfacedEventHandler. InterfacedEventHandler accepts anything that implements EventListener.

**Comment by mentlerd** [comment]
Apologies, my wording is really not clear on that one.

I am aware how ListenerAdapter is supposed to be used, what I meant is that ListenerAdapter forces your listener objects to extend from it, preventing you from having your own class hierarchy, while it does not really have to anymore, since you can implement the same functionality with interface defaults.

In my example usecase, my bot is built from modules, that require to extend a common superclass, for configuration/lifecycle management. I am looking for a way for the module object to receive events, without having to either create a delegating ListenerAdapter subclass, or doing instanceof checks.

**Comment by natanbc** [comment]
You could use the AnnotatedEventManager

**Comment by MinnDevelopment** [comment]
If you wish for a different event system you should implement your own IEventManager and event listener system.

**Comment by mentlerd** [comment]
I was not aware there were more than one IEventManager implementations available. Thanks!
Probably going to end up making my own implementation anyway..

My initial point was that a visitor system cannot be created without cooperation from the visited objects, so I can't make a system like I have described in the first post without modifying core files.

I am not really asking for someone to code this for me, rather if I made the change as described, would it get merged or not?
